### PR TITLE
A5: Selection Sort and Binary Search

### PR DESCRIPTION
**Pull request by sudarson-25**
PR created for review

### CODE REVIEW

File: Training/Training/Training.c [hunk 1]
```diff
@@ -1,14 +1,226 @@
...
-
-int main () {
-	printf ("Hello World!");
-	return 0;
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// Training.c
+// Program to to perform Selection Sort and Binary Search
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]:
```suggestion
// Implements selection sort and binary search

```

File: Training/Training/Training.c [hunk 2]
```diff
@@ -1,14 +1,226 @@
...
+// Sudarson S
+// ------------------------------------------------------------------------------------------------
+
+#include<stdio.h>
+#include<stdbool.h>
+#include<errno.h>
+#include <string.h>
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]:
```suggestion
#include<string.h>

```

File: Training/Training/Training.c [hunk 3]
```diff
@@ -1,14 +1,226 @@
...
+#include<malloc.h>
+#include<stdlib.h>
+
+#define column 8
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Instead of defining as constant, use 8 directly.

File: Training/Training/Training.c [hunk 4]
```diff
@@ -1,14 +1,226 @@
...
+#define ANSI_COLOR_RED     "\x1b[31m"
+#define ANSI_COLOR_GREEN   "\x1b[32m"
+#define ANSI_COLOR_YELLOW  "\x1b[33m"
+#define ANSI_COLOR_RESET   "\x1b[0m"
+
+/// <summary>Function implementing selection sort logic </summary>
+void SelectionSort (int* arr, int n) {
+   for (int i = 0; i < n - 1; i++) {
+      int minInd = i;
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]:
```suggestion
      int min_idx = i;

```

File: Training/Training/Training.c [hunk 5]
```diff
@@ -1,14 +1,226 @@
...
+      // Iterate through the unsorted portion to find the actual minimum
+      for (int j = i + 1; j < n; j++)
+         if (arr[j] < arr[minInd])
+            // Update minInd if a smaller element is found
+            minInd = j;
+      // If a new minimum is found, swap it with the element at index i
+      if (minInd != i) {
+         int temp = arr[i];
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Try to swap without using temporary variable.

File: Training/Training/Training.c [hunk 6]
```diff
@@ -1,14 +1,226 @@
...
+#include<malloc.h>
+#include<stdlib.h>
+
+#define column 8
+#define ANSI_COLOR_RED     "\x1b[31m"
+#define ANSI_COLOR_GREEN   "\x1b[32m"
+#define ANSI_COLOR_YELLOW  "\x1b[33m"
+#define ANSI_COLOR_RESET   "\x1b[0m"
+
+/// <summary>Function implementing selection sort logic </summary>
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Not necessary to mention 'function' as it is known. Apply this wherever possible.
```suggestion
/// <summary>Implements selection sort algorithm</summary>

```

File: Training/Training/Training.c [hunk 7]
```diff
@@ -1,14 +1,226 @@
...
+   errno = 0;
+   char* endptr = NULL;
+   *num = strtol (number, &endptr, 10);
+   return endptr == number || *endptr != '\n' || errno == ERANGE || number[0] == ' ' ||
+      number[0] == '\t' ?
+      false : true;
+}
+
+/// <summary>Function to drain the unassigned inputs from stdin</summary>
+void BufferDrain () {
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Avoid a separate function for a single line.

File: Training/Training/Training.c [hunk 8]
```diff
@@ -1,14 +1,226 @@
...
+   char choice[3], * result = NULL;
+   while (1) {
+      printf ("Menu\n~~~~\n1. Selection Sort\n2. Binary Search\n3. Run Test Cases\n"
+              "4. Exit\n\nEnter your choice: ");
+      fgets (choice, sizeof (choice), stdin);
+      result = strchr (choice, '\n');
+      if (result == NULL) {
+         printf ("Invalid Choice!\n");
+         BufferDrain ();
+      } else
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]:
```suggestion
      if (result == NULL) {
         printf ("Invalid Choice!\n");
         BufferDrain ();
         continue;
      } 

```

File: Training/Training/Training.c [hunk 9]
```diff
@@ -1,14 +1,226 @@
...
+      false : true;
+}
+
+/// <summary>Function to drain the unassigned inputs from stdin</summary>
+void BufferDrain () {
+   while (getchar () != '\n');
+}
+
+/// <summary>Function to accept user inputs</summary>
+void UserInputs (int** arr, int* arrSize) {
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Please cut short this user input function.

File: Training/Training/Training.c [hunk 10]
```diff
@@ -1,14 +1,226 @@
...
+                  printf ("Unsorted Array: ");
+                  printArray (arr, arrSize);
+                  SelectionSort (arr, arrSize);
+                  printf ("\nSorted Array: ");
+                  printArray (arr, arrSize);
+                  free (arr);
+               }break;
+            case '2': {
+                  char keyStr[13];
+                  bool keyFlag;
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Why is this variable needed?

**Review by ShakthyTMM — CHANGES_REQUESTED** [review]
Have separate files for headers and test. Your Training.c should only contain selection sort and binary search functions.

File: Training/Training/header.c [hunk 11]
```diff
@@ -0,0 +1,34 @@
+#include "header.h"
```
Comment (ShakthyTMM) [inline Training/Training/header.c]: Add program headers to all the files.

File: Training/Training/Training.c [hunk 12]
```diff
@@ -1,14 +1,138 @@
...
-
-int main () {
-	printf ("Hello World!");
-	return 0;
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// Training.c
+// Implements selection sort and binary search
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Please change the header according to each file.

File: Training/Training/Training.c [hunk 13]
```diff
@@ -1,14 +1,138 @@
...
+/// <summary>Accepts user inputs</summary>
+void UserInputs (int** arr, int* arrSize) {
+   printf ("\n");
+   do {
+      IsValidInteger ("Enter an array size: ", arrSize);
+      if (*arrSize <= 0)
+         printf ("Invalid Input\n\n");
+   } while (*arrSize <= 0);
+   do {
+      *arr = (int*)malloc ((*arrSize) * sizeof (int));
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Remove the parenthesis as it only one statement.

File: Training/Training/Training.c [hunk 14]
```diff
@@ -1,14 +1,138 @@
...
+      if (output == false)
+         printf ("Invalid Input\n\n");
+   } while (output == false);
+}
+
+/// <summary>Accepts user inputs</summary>
+void UserInputs (int** arr, int* arrSize) {
+   printf ("\n");
+   do {
+      IsValidInteger ("Enter an array size: ", arrSize);
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Apply this wherever possible.
```suggestion
   do {
      IsValidInteger ("\nEnter an array size: ", arrSize);

```

File: Training/Training/Training.c [hunk 15]
```diff
@@ -1,14 +1,138 @@
...
+      printArray (sortExpected[i], 8);
+      SelectionSort (inputs[i], 8);
+      printf ("\nSorted:   ");
+      printArray (inputs[i], 8);
+      for (int j = 0; j < 8; j++)
+         if (inputs[i][j] == sortExpected[i][j])
+            flag++;
+      flag == 8 ?
+         printf ("\nResult:   "ANSI_COLOR_GREEN"Passed\n"ANSI_COLOR_RESET) :
+         printf ("\nResult:   "ANSI_COLOR_RED"Failed\n"ANSI_COLOR_RESET);
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Why to have a flag? If the values don't match straight away print 'failed' and break the loop. To check for passing cases, use the iterating variable 'j'.

File: Training/Training/header.c [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+         arr[minIdx] = arr[i] - arr[minIdx];
+         arr[i] -= arr[minIdx];
+      }
+   }
+}
+
+int BinarySearch (int* arr, int arrSize, int key) {
+   int low = 0, high = arrSize - 1, idx = -1;
+   while (low <= high) {
+      int mid = low + (high - low) / 2;
```
Comment (ShakthyTMM) [inline Training/Training/header.c]: Can't this be high+low/2?

File: Training/Training/Training.c [hunk 17]
```diff
@@ -1,14 +1,138 @@
...
-	return 0;
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// Training.c
+// Implements selection sort and binary search
+// Sudarson S
+// ------------------------------------------------------------------------------------------------
+
```
Comment (ShakthyTMM) [inline Training/Training/Training.c:10]: Remove the extra line.

**Review by ShakthyTMM — CHANGES_REQUESTED** [review]
Instead of having selection sort and binary search as different options get an array and sort it by selection sort algorithm and ask for which element to search in that sorted array.

File: Training/Training/header.c (hunk 16 above)
Comment (sudarson-25) [reply Training/Training/header.c:23]: It would work only for low = 0, if low had been changed it wouldn't work.

File: Training/Training/header.c
Comment (ShakthyTMM) [inline Training/Training/header.c:1]: Change the name of this file to training.c and test file to test.c

File: Training/Training/Training.c [hunk 18]
```diff
@@ -1,14 +1,123 @@
...
-
-int main () {
-	printf ("Hello World!");
-	return 0;
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// Training.c
+// Executes the main function
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]:
```suggestion
// Tests for selection sort and binary search

```

File: Training/Training/Training.c
Comment (ShakthyTMM) [inline Training/Training/Training.c:1]: Please arrange the functions in a file in ascending order. Also please try to practice it here on.

File: Training/Training/Training.c [hunk 19]
```diff
@@ -1,14 +1,123 @@
...
+}
+
+/// <summary>Runs test cases</summary>
+void TestCases () {
+   int inputs[][8] = { {0,12,3,4,4,4,4,4},{-98,45,2,1,3,4,0,0},{6,7,8,9,-4,-5,-9,0},{0,0,0,0,0,0,0,0} }, j,
+      binSearchKey[] = { 12,4,-4,6 }, binExpected[] = { 7,6,2,-1 }, n = sizeof (inputs) / sizeof (inputs[0]),
+      sortExpected[][8] = { { 0,3,4,4,4,4,4,12},{-98,0,0,1,2,3,4,45 },{-9,-5,-4,0,6,7,8,9 },{0,0,0,0,0,0,0,0} };
+   printf ("\nTestCases\n~~~~~~~~~");
+   for (int i = 0; i < n; i++) {
+      int output, flag = 0;
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Before commenting, please check if you have removed the unnecessary variables after making a change.

File: Training/Training/header.h [hunk 20]
```diff
@@ -0,0 +1,20 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// header.h
+// Contains function signatures
+// Sudarson S
+// ------------------------------------------------------------------------------------------------
+#pragma once
```
Comment (ShakthyTMM) [inline Training/Training/header.h]: Always keep this as the first line of the file.

**Review by ShakthyTMM — CHANGES_REQUESTED** [review]

File: Training/Training/Training.c [hunk 21]
```diff
@@ -1,14 +1,42 @@
...
+      int mid = (high + low) / 2;
+      // Checks if key is present at mid and is it's first occurrence
+      if (arr[mid] == key) {
+         idx = mid;
+         high = mid - 1;
+      } else if (arr[mid] < key)
+         low = mid + 1;
+      else
+         high = mid - 1;
+   }
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Always try to use ternary operator instead of if else block. This will improve your code elegance.
```suggestion
            (arr[mid] == key) ? (idx = mid, high = mid - 1) : (arr[mid] < key ? (low = mid + 1) : (high = mid - 1));

```

**Review by ShakthyTMM — CHANGES_REQUESTED** [review]

File: Training/Training/test.c [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+         *num = strtol (numberStr, &endptr, 10);
+         output = endptr == numberStr || *endptr != '\n' || errno == ERANGE || numberStr[0] == ' ' ||
+            numberStr[0] == '\t' ? false : true;
+      }
+      if (output == false) printf ("Invalid Input\n");
+   } while (output == false);
+}
+
+/// <summary>Prints the array elements</summary>
+void PrintArray (int* arr, int n) {
```
Comment (ShakthyTMM) [inline Training/Training/test.c]: Please don't write a function for a single line.

**Review by ShakthyTMM — CHANGES_REQUESTED** [review]

**Review by ShakthyTMM — COMMENTED** [review]
Screenshot 2024-10-25 101649

Why the first one is invalid while the second is valid?

Also, you have fixed the number of inputs for the test cases to 8. What if I want to add few more to test? This will cause an out of range exception.
Always try to put yourself in the place of a tester and try hard to find bugs. This way you can frame a good unit test.

File: Training/Training/Training.c [hunk 23]
```diff
@@ -1,14 +1,34 @@
...
+// Sudarson S
+// ------------------------------------------------------------------------------------------------
+#include"training.h"
+
+int BinarySearch (int* arr, int arrSize, int key) {
+   int low = 0, high = arrSize - 1, idx = -1;
+   while (low <= high) {
+      int mid = low + (high - low) / 2;
+      // Checks if key is present at mid and is it's first occurrence
+      arr[mid] == key ? idx = mid, high = mid - 1 : arr[mid] < key ? low = mid + 1 : (high = mid - 1);
```
Comment (ShakthyTMM) [inline Training/Training/Training.c]: Mostly ternary operator evaluates a Boolean expression and returns either one expression. Here it is mere assignment. So, 'if else' can be used.

File: Training/Training/test.c [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+
+/// <summary>Runs test cases for the sorting algorithm</summary>
+void SortTestCases () {
+   int* jagged[8] = { 0 }, size[8] = { 0 };
+   srand ((unsigned int)time (NULL));
+   for (int i = 0; i < 8; i++) {
+      size[i] = rand () % (20 - 1 + 1) + 1; // Range [1,20]
+      jagged[i] = malloc (sizeof (int) * size[i]);
+      for (int j = 0; j < size[i]; j++) {
+         jagged[i][j] = rand () % (1000 + 1000 + 1) - 1000; // Range [-1000,1000]
```
Comment (ShakthyTMM) [inline Training/Training/test.c]: The range can be [INT_MIN, INT_MAX].

**Review by ShakthyTMM — CHANGES_REQUESTED** [review]

File: Training/Training/test.c [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+         while (getchar () != '\n');
+         output = false;
+      } else {
+         errno = 0;
+         char* endptr = NULL;
+         *num = strtol (numberStr, &endptr, 10);
+         output = endptr != numberStr && *endptr == '\n' && errno != ERANGE && numberStr[0] != ' ' &&
+            numberStr[0] != '\t';
+      }
+      if (output == false) printf ("Invalid Input\n");
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
      if (!output) printf ("Invalid Input\n");

```

File: Training/Training/test.c [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+   } while (output == false);
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
   } while (!output);

```

File: Training/Training/test.c [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+#include<stdbool.h>
+#include<stdio.h>
+#include<string.h> // USED
+#include<errno.h>
+#include<malloc.h>
+#include<stdlib.h>
+#include <time.h>
+
+/// <summary>Checks if a given input is a valid integer</summary>
+void IsValidInteger (char* prompt, int* num) {
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]: By convention when a function name start with 'Is' we expect the function to return a boolean value. Here, you assign a value to output, but you don't return it. The better way here is to return instead of printing.

File: Training/Training/test.c [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+            numberStr[0] != '\t';
+      }
+      if (output == false) printf ("Invalid Input\n");
+   } while (output == false);
+}
+
+/// <summary>Runs test cases for the sorting algorithm</summary>
+void SortTestCases () {
+   srand ((unsigned int)time (NULL));
+   // Range of n [1,20]
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]: Here it would be helpful to have comments describing what you are attempting to do (Create random arrays of random length to test your algorithm).
```suggestion


```

File: Training/Training/test.c [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+   int n = rand () % (20 - 1 + 1) + 1, ** jagged = malloc (n * sizeof (int*)),
+      * size = malloc (n * sizeof (int));
+   for (int i = 0; i < n; i++) {
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c:44]: I'd like for you to stick to the previous test output format.

File: Training/Training/test.c [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+      size[i] = rand () % (20 - 1 + 1) + 1; // Range [1,20]
+      jagged[i] = malloc (sizeof (int) * size[i]);
+      for (int j = 0; j < size[i]; j++) {
+         jagged[i][j] = rand () % (10000 + 10000 + 1) - 10000; // Range [-10000,10000]
+      }
+   }
+   printf ("\nTestCases\n~~~~~~~~~\nSelection Sort\n~~~~~~~~~~~~~");
+   for (int i = 0; i < n; i++) {
+      int k;
+      printf (ANSI_COLOR_YELLOW"\n----------------------------------------------------------------"
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
      char* dashes = ANSI_COLOR_YELLOW"\n----------------------------------------------------------------"
   "-----------------------------------------"ANSI_COLOR_RESET;
printf ("%s\nInput:    ", dashes);

```
You can re-use dashes this way.

File: Training/Training/test.c [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+      SelectionSort (jagged[i], size[i]);
+      printf ("\nSorted:   ");
+      for (int j = 0; j < size[i]; j++) printf ("%d ", jagged[i][j]);
+      for (k = 0; k < size[i] - 1; k++)
+         if (jagged[i][k + 1] < jagged[i][k]) {
+            printf ("\nResult:   "ANSI_COLOR_RED"Failed\n"ANSI_COLOR_YELLOW"----------------------"
+               "-----------------------------------------------------------------------------------"
+               ANSI_COLOR_RESET);
+            break;
+         }
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
   for (k = 0; jagged[i][k + 1] > jagged[i][k]; k++);


```

File: Training/Training/test.c [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+      if (k == size[i] - 1) printf ("\nResult:   "ANSI_COLOR_GREEN"Passed\n"ANSI_COLOR_YELLOW"----"
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]: Your for loop is designed only to break when the condition fails (so it will only reach  k = size -1 if the condition is met every time. You can use this to assign result.
```suggestion
      char* result = k == size[i] - 1 ? ANSI_COLOR_GREEN"Passed" : ANSI_COLOR_RED"Failed";
      printf ("\nResult:   %s\n%s", result, dashes);

```

File: Training/Training/test.c [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+         "----------------------------------------------------------------------------------------"
+         "-------------"ANSI_COLOR_RESET);
+   }
+   free (jagged);
+   free (size);
+}
+
+/// <summary>Runs test cases for the searching algorithm</summary>
+void SearchTestCases () {
+   int row0[] = { 20,19,-5,30,21,12 }, row1[] = { -5, 6 }, row2[] = { 0,-12,3,4,-4,4,-4,4 }, row3[] = { -250, 60, 46, 17, 15 },
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]: We try to limit max chars in a row to 100 for improving readability. 
```suggestion
   int row0[] = { 20,19,-5,30,21,12 }, row1[] = { -5, 6 }, row2[] = { 0, -12, 3, 4, -4, 4, -4, 4 },
    row3[] = { -250, 60, 46, 17, 15 },

```

File: Training/Training/test.c [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+      * jagged[] = { row0, row1, row2, row3 }, size[] = { 6, 2, 8, 5 }, binSearchKey[] = { 20, 6, 4, 17 },
+      binExpected[] = { 3, 1, 5, 2 }, output;
+   printf ("\nTestCases\n~~~~~~~~~\nBinary Search\n~~~~~~~~~~~~~");
+   for (int i = 0; i < 4; i++) {
+      printf (ANSI_COLOR_YELLOW"\n----------------------------------------------------------------"
+         "-----------------------------------------"ANSI_COLOR_RESET"\nInput:          ");
+      for (int j = 0; j < size[i]; j++) printf ("%d ", jagged[i][j]);
+      SelectionSort (jagged[i], size[i]);
+      printf ("\nSorted:         ");
+      for (int j = 0; j < size[i]; j++) printf ("%d ", jagged[i][j]);
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]: There is no need to print the sort and print sorted arr for search cases. Binary search expects a sorted array and for test cases you may simply input sorted arrays. To indicate you expect sorted arrays to the user, you may rename the formal parameter to sortedArr

File: Training/Training/test.c [hunk 35]
```diff
@@ -0,0 +1,144 @@
...
+         "   %d", binSearchKey[i], binExpected[i], output = BinarySearch (jagged[i], size[i], binSearchKey[i]));
+      printf (output == binExpected[i] ? "\nResult:         "ANSI_COLOR_GREEN"Passed\n"ANSI_COLOR_RESET :
+         "\nResult:         "ANSI_COLOR_RED"Failed\n"ANSI_COLOR_RESET);
+      printf (ANSI_COLOR_YELLOW"------------------------------------------------------------------"
+         "---------------------------------------"ANSI_COLOR_RESET);
+   }
+}
+
+/// <summary>Accepts user inputs</summary>
+void UserInputs (int** arr, int* arrSize) {
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]: image
Please try to make changes suggested by intellisense wherever appropriate.

File: Training/Training/test.c [hunk 36]
```diff
@@ -0,0 +1,144 @@
...
+#include<stdio.h>
+#include<string.h> // USED
+#include<errno.h>
+#include<malloc.h>
+#include<stdlib.h>
+#include <time.h>
+
+/// <summary>Checks if a given input is a valid integer</summary>
+void IsValidInteger (char* prompt, int* num) {
+   bool output;
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c:21]: Since this is actually being used to "Get" user input, it would be more appropriate to call it GetValidInteger

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

**Comment by Dakshita-Minocha** [comment]
Please start naming your branches as A5, A6 and so on

File: Training/Training/test.c [hunk 37]
```diff
@@ -0,0 +1,132 @@
...
+         output = false;
+      } else {
+         errno = 0;
+         char* endptr = NULL;
+         num = strtol (numberStr, &endptr, 10);
+         output = endptr != numberStr && *endptr == '\n' && errno != ERANGE && numberStr[0] != ' ' &&
+            numberStr[0] != '\t';
+      }
+      if (!output) printf ("Invalid Input\n");
+   } while (output == false);
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
   } while (!output);

```
Always check your code to see if there are other places certain changes are applicable.

File: Training/Training/test.c [hunk 38]
```diff
@@ -0,0 +1,132 @@
...
+   bool output;
+   int num;
+   do {
+      char numberStr[4096], * result = NULL;
+      printf ("%s", prompt);
+      fgets (numberStr, sizeof (numberStr), stdin);
+      result = strchr (numberStr, '\n');
+      if (result == NULL) {
+         while (getchar () != '\n');
+         output = false;
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
         printf ("Invalid input");
         continue;

```

File: Training/Training/test.c [hunk 39]
```diff
@@ -0,0 +1,132 @@
...
+      } else {
+         errno = 0;
+         char* endptr = NULL;
+         num = strtol (numberStr, &endptr, 10);
+         output = endptr != numberStr && *endptr == '\n' && errno != ERANGE && numberStr[0] != ' ' &&
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
         output = endptr != NULL && num != 0;

```

File: Training/Training/test.c [hunk 40]
```diff
@@ -0,0 +1,132 @@
...
+            numberStr[0] != '\t';
+      }
+      if (!output) printf ("Invalid Input\n");
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c:33]:
```suggestion

```

File: Training/Training/test.c [hunk 41]
```diff
@@ -0,0 +1,132 @@
...
+      printf ("\nResult:   %s\n%s", result, dashes);
+   }
+   free (jagged);
+   free (size);
+}
+
+/// <summary>Runs test cases for the searching algorithm</summary>
+static void SearchTestCases () {
+   int row0[] = { -5, 12, 19, 20, 21, 30 }, row1[] = { -5, 6 }, row2[] = { -12, -4, -4, 0, 3, 4, 4,
+      4 }, row3[] = { -250, 15, 17, 46, 60 }, * jagged[] = { row0, row1, row2, row3 }, size[] = { 6,
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
   int row0[] = { -5, 12, 19, 20, 21, 30 }, row1[] = { -5, 6 }, row2[] = { -12, -4, -4, 0, 3, 4, 4, 4 },
       row3[] = { -250, 15, 17, 46, 60 }, * jagged[] = { row0, row1, row2, row3 }, size[] = { 6,

```

File: Training/Training/test.c [hunk 42]
```diff
@@ -0,0 +1,132 @@
...
+               arrSize = UserInputs (&arr);
+               printf ("\nUnsorted Array: ");
+               for (int i = 0; i < arrSize; i++) printf ("%d ", arr[i]);
+               SelectionSort (arr, arrSize);
+               printf ("\nSorted Array: ");
+               for (int i = 0; i < arrSize; i++) printf ("%d ", arr[i]);
+               key = GetValidInteger ("\nEnter the element to be found: ");
+               printf ("Element found at index: %d", BinarySearch (arr, arrSize, key));
+               free (arr);
+            }break;
```
Comment (Dakshita-Minocha) [inline Training/Training/test.c]:
```suggestion
            } break;

```

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

**Review by Dakshita-Minocha — APPROVED** [review]
